tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -226,16 +232,277 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
+        jsClassFieldDefsAskers -= invalidatable
+      }
+
+      /** Call this when we invalidate all caches. */
+      def forgetAllAskers(): Unit = {
+        isInterfaceAskers.clear()
+        jsNameAskers.clear()
+        jsSuperClassAskers.clear()
+        jsClassFieldDefsAskers.clear()
+      }
+    }
+
+    private val classes = mutable.Map.empty[String, Class]
+
+    /** Returns `true` if *all* caches should be invalidated.
+     *
+     *  For global properties that are rarely changed and heavily used (such as
+     *  isParentDataAccessed), we do not want to pay the price of the
+     *  dependency graph, in terms of memory consumption and time spent
+     *  maintaining it. It is a better trade-off to invalidate everything in
+     *  the rare events where they do change.
+     */
+    def startRun(linkingUnit: LinkingUnit): Boolean = {
+      val newIsParentDataAccessed = linkingUnit.globalInfo.isParentDataAccessed
+      var newHasNewRuntimeLong: Boolean = false
+
+      // Update classes
+      for (linkedClass <- linkingUnit.classDefs) {
+        classes.get(linkedClass.encodedName).fold[Unit] {
+          // new class
+          classes.put(linkedClass.encodedName, new Class(linkedClass))
+        } { existingCls =>
+          existingCls.update(linkedClass)
+        }
+
+        if (linkedClass.encodedName == LongImpl.RuntimeLongClass) {
+          newHasNewRuntimeLong = linkedClass.memberMethods.exists { linkedMethod =>
+            linkedMethod.tree.name.name == LongImpl.initFromParts
+          }
+        }
+      }
+
+      // Garbage collection
+      classes.retain((_, cls) => cls.testAndResetIsAlive())
+
+      val invalidateAll = !firstRun && {
+        newIsParentDataAccessed != isParentDataAccessed ||
+        newHasNewRuntimeLong != hasNewRuntimeLong
+      }
+      firstRun = false
+
+      isParentDataAccessed = newIsParentDataAccessed
+      hasNewRuntimeLong = newHasNewRuntimeLong
+
+      if (invalidateAll)
+        classes.valuesIterator.foreach(_.forgetAllAskers())
+      invalidateAll
+    }
+
+    def askHasNewRuntimeLong(invalidatable: Invalidatable): Boolean =
+      hasNewRuntimeLong
+
+    def askIsParentDataAccessed(invalidatable: Invalidatable): Boolean =
+      isParentDataAccessed
+
+    def askIsInterface(className: String, invalidatable: Invalidatable): Boolean =
+      classes(className).askIsInterface(invalidatable)
+
+    def askJSClassJSName(className: String, invalidatable: Invalidatable): Option[String] =
+      classes(className).askJSClassJSName(invalidatable)
+
+    def askJSSuperClass(className: String, invalidatable: Invalidatable): String =
+      classes(className).askJSSuperClass(invalidatable)
+
+    def askJSClassFieldDefs(className: String, invalidatable: Invalidatable): List[FieldDef] =
+      classes(className).askJSClassFieldDefs(invalidatable)
+  }
+
   // Caching
 
-  private final class ClassCache {
+  private abstract class Invalidatable extends GlobalKnowledge {
```

Comment:
I would reserve `Invalidatable` for really just the basic invalidatable (with `invalidate()` and nothing more).

Maybe this is a `KnowledgeAccessor` or something?